// react-rainbow-components/components/Tab/pageObject/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

/**
 * Tab page object class.
 * @class
 */
var PageTab =
/*#__PURE__*/
function () {
  /**
   * Create a new Tab page object.
   * @constructor
   * @param {string} rootElement - The selector of the Tab root element.
   */
  function PageTab(rootElement) {
    (0, _classCallCheck2["default"])(this, PageTab);
    this.rootElement = rootElement;
  }
  /**
   * Clicks the tab item.
   * @method
   */


  (0, _createClass2["default"])(PageTab, [{
    key: "click",
    value: function click() {
      $(this.rootElement).$('a').click();
    }
    /**
     * Returns true when the tab item has focus.
     * @method
     * @returns {bool}
     */

  }, {
    key: "hasFocus",
    value: function hasFocus() {
      return $(this.rootElement).$('a').isFocused();
    }
    /**
     * Returns true when the tab item is selected.
     * @method
     * @returns {bool}
     */

  }, {
    key: "isSelected",
    value: function isSelected() {
      return $(this.rootElement).$('a').getAttribute('class') === 'rainbow-tab_anchor rainbow-tab--active';
    }
    /**
     * Returns true when the tab item is visible in the viewport.
     * @method
     * @returns {bool}
     */

  }, {
    key: "isVisibleWithinViewport",
    value: function isVisibleWithinViewport() {
      return $(this.rootElement).$('a').isDisplayedInViewport();
    }
  }]);
  return PageTab;
}();

module.exports = PageTab;
