// react-rainbow-components/components/Picklist/label.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Label;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _RequiredAsterisk = _interopRequireDefault(require("../RequiredAsterisk"));

function Label(props) {
  var label = props.label,
      required = props.required,
      inputId = props.inputId,
      id = props.id;

  var getLabelClassNames = function getLabelClassNames() {
    var readOnly = props.readOnly,
        hideLabel = props.hideLabel;
    return (0, _classnames["default"])('rainbow-lookup_input-label', {
      'rainbow-lookup_input-label-read-only': readOnly,
      'rainbow-picklist_input-label--hide': hideLabel
    });
  };

  return _react["default"].createElement("label", {
    className: getLabelClassNames(),
    htmlFor: inputId,
    id: id
  }, _react["default"].createElement(_RequiredAsterisk["default"], {
    required: required
  }), label);
}

Label.propTypes = {
  label: _propTypes["default"].node,
  required: _propTypes["default"].bool,
  inputId: _propTypes["default"].string,
  readOnly: _propTypes["default"].bool,
  id: _propTypes["default"].string,
  hideLabel: _propTypes["default"].bool
};
Label.defaultProps = {
  required: false,
  inputId: undefined,
  readOnly: false,
  id: undefined,
  label: null,
  hideLabel: false
};
