// react-rainbow-components/components/RadioGroup/radio.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _RenderIf = _interopRequireDefault(require("../RenderIf"));

var _utils = require("../../libs/utils");

var Radio =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(Radio, _Component);

  function Radio(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, Radio);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Radio).call(this, props));
    _this.radioId = (0, _utils.uniqueId)('radio');
    return _this;
  }

  (0, _createClass2["default"])(Radio, [{
    key: "getLabelClassNames",
    value: function getLabelClassNames() {
      var disabled = this.props.disabled;
      return (0, _classnames["default"])('rainbow-radio-group_radio-label', {
        'rainbow-radio-group_radio-label--disabled': disabled
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          label = _this$props.label,
          ariaDescribedby = _this$props.ariaDescribedby,
          onChange = _this$props.onChange,
          value = _this$props.value,
          disabled = _this$props.disabled,
          isChecked = _this$props.isChecked,
          name = _this$props.name,
          required = _this$props.required;
      return _react["default"].createElement("span", {
        className: "rainbow-radio-group_radio"
      }, _react["default"].createElement("input", {
        type: "radio",
        required: required,
        id: this.radioId,
        name: name,
        value: value,
        checked: isChecked,
        "aria-describedby": ariaDescribedby,
        onChange: onChange,
        disabled: disabled
      }), _react["default"].createElement("label", {
        className: "rainbow-radio-group_radio-label-container",
        htmlFor: this.radioId
      }, _react["default"].createElement("span", {
        className: "rainbow-radio-group_radio-faux"
      }), _react["default"].createElement(_RenderIf["default"], {
        isTrue: !!label
      }, _react["default"].createElement("span", {
        className: this.getLabelClassNames()
      }, label))));
    }
  }]);
  return Radio;
}(_react.Component);

exports["default"] = Radio;
Radio.propTypes = {
  label: _propTypes["default"].node.isRequired,
  value: _propTypes["default"].string,
  onChange: _propTypes["default"].func,
  ariaDescribedby: _propTypes["default"].string,
  disabled: _propTypes["default"].bool,
  isChecked: _propTypes["default"].bool,
  name: _propTypes["default"].string,
  required: _propTypes["default"].bool.isRequired
};
Radio.defaultProps = {
  value: undefined,
  onChange: function onChange() {},
  ariaDescribedby: undefined,
  disabled: false,
  isChecked: false,
  name: undefined
};
