// react-rainbow-components/components/RadioButtonGroup/radioButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _utils = require("../../libs/utils");

var RadioButton =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(RadioButton, _Component);

  function RadioButton(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, RadioButton);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(RadioButton).call(this, props));
    _this.radioId = (0, _utils.uniqueId)('radiobutton');
    return _this;
  }

  (0, _createClass2["default"])(RadioButton, [{
    key: "getContainerClassNames",
    value: function getContainerClassNames() {
      var _this$props = this.props,
          isChecked = _this$props.isChecked,
          disabled = _this$props.disabled;
      return (0, _classnames["default"])('rainbow-radio-button-group_radio', {
        'rainbow-radio-button-group_radio--checked': isChecked,
        'rainbow-radio-button-group_radio--disabled': disabled
      });
    }
  }, {
    key: "getLabelClassNames",
    value: function getLabelClassNames() {
      var disabled = this.props.disabled;
      return (0, _classnames["default"])('rainbow-radio-button-group_radio-label', {
        'rainbow-radio-button-group_radio-label--disabled': disabled
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          itemRef = _this$props2.itemRef,
          label = _this$props2.label,
          ariaDescribedby = _this$props2.ariaDescribedby,
          onChange = _this$props2.onChange,
          value = _this$props2.value,
          disabled = _this$props2.disabled,
          isChecked = _this$props2.isChecked,
          name = _this$props2.name,
          required = _this$props2.required;
      return _react["default"].createElement("span", {
        ref: itemRef,
        className: this.getContainerClassNames()
      }, _react["default"].createElement("input", {
        type: "radio",
        required: required,
        id: this.radioId,
        name: name,
        value: value,
        checked: isChecked,
        "aria-describedby": ariaDescribedby,
        onChange: onChange,
        disabled: disabled
      }), _react["default"].createElement("label", {
        className: "rainbow-radio-button-group_radio-label-container",
        htmlFor: this.radioId
      }, _react["default"].createElement("span", {
        className: this.getLabelClassNames()
      }, label)));
    }
  }]);
  return RadioButton;
}(_react.Component);

exports["default"] = RadioButton;
RadioButton.propTypes = {
  label: _propTypes["default"].node.isRequired,
  value: _propTypes["default"].string,
  onChange: _propTypes["default"].func,
  ariaDescribedby: _propTypes["default"].string,
  disabled: _propTypes["default"].bool,
  isChecked: _propTypes["default"].bool,
  name: _propTypes["default"].string,
  required: _propTypes["default"].bool.isRequired,
  itemRef: _propTypes["default"].object.isRequired
};
RadioButton.defaultProps = {
  value: undefined,
  onChange: function onChange() {},
  ariaDescribedby: undefined,
  disabled: false,
  isChecked: false,
  name: undefined
};
