// react-rainbow-components/components/TimePicker/timeSelect.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ButtonIcon = _interopRequireDefault(require("../ButtonIcon"));

var _Button = _interopRequireDefault(require("../Button"));

var _ampmSelect = _interopRequireDefault(require("./ampmSelect"));

var _upArrow = _interopRequireDefault(require("./icons/upArrow"));

var _downArrow = _interopRequireDefault(require("./icons/downArrow"));

var _RenderIf = _interopRequireDefault(require("../RenderIf"));

var _normalizeHour = _interopRequireDefault(require("./helpers/normalizeHour"));

var _normalizeMinutes = _interopRequireDefault(require("./helpers/normalizeMinutes"));

var _getNextHour = _interopRequireDefault(require("./helpers/getNextHour"));

var _getPrevHour = _interopRequireDefault(require("./helpers/getPrevHour"));

var _getNextMinute = _interopRequireDefault(require("./helpers/getNextMinute"));

var _getPrevMinute = _interopRequireDefault(require("./helpers/getPrevMinute"));

var _getNextAmPmValue = _interopRequireDefault(require("./helpers/getNextAmPmValue"));

var _get24HourTime = _interopRequireDefault(require("./helpers/get24HourTime"));

var _getSingleNewTypedValue = _interopRequireDefault(require("./helpers/getSingleNewTypedValue"));

var _isNumber = _interopRequireDefault(require("./helpers/isNumber"));

var _getHour = _interopRequireDefault(require("./helpers/getHour"));

var _getMinutes = _interopRequireDefault(require("./helpers/getMinutes"));

var _getAmPm = _interopRequireDefault(require("./helpers/getAmPm"));

var _getDefaultAmPm = _interopRequireDefault(require("./helpers/getDefaultAmPm"));

var _constants = require("../../libs/constants");

function preventDefault(event) {
  event.preventDefault();
}

var TimeSelect =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(TimeSelect, _Component);

  function TimeSelect(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, TimeSelect);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(TimeSelect).call(this, props));
    _this.state = {
      hour: (0, _getHour["default"])(props.value),
      minutes: (0, _getMinutes["default"])(props.value),
      ampm: (0, _getAmPm["default"])(props.value)
    };
    _this.hourInputRef = _react["default"].createRef();
    _this.minutesInputRef = _react["default"].createRef();
    _this.amPmInputRef = _react["default"].createRef();
    _this.inputsMap = {
      0: _this.hourInputRef,
      1: _this.minutesInputRef,
      2: _this.amPmInputRef
    };
    _this.inputFocusedIndex = 0;
    _this.hasPropValue = !!props.value;
    _this.handleChangeHour = _this.handleChangeHour.bind((0, _assertThisInitialized2["default"])(_this));
    _this.handleFocusHour = _this.handleFocusHour.bind((0, _assertThisInitialized2["default"])(_this));
    _this.handleBlurHour = _this.handleBlurHour.bind((0, _assertThisInitialized2["default"])(_this));
    _this.handleChangeMinutes = _this.handleChangeMinutes.bind((0, _assertThisInitialized2["default"])(_this));
    _this.handleFocusMinutes = _this.handleFocusMinutes.bind((0, _assertThisInitialized2["default"])(_this));
    _this.handleAmPmChange = _this.handleAmPmChange.bind((0, _assertThisInitialized2["default"])(_this));
    _this.hanldeFocusAmPm = _this.hanldeFocusAmPm.bind((0, _assertThisInitialized2["default"])(_this));
    _this.handleKeyDown = _this.handleKeyDown.bind((0, _assertThisInitialized2["default"])(_this));
    _this.incrementHandler = _this.incrementHandler.bind((0, _assertThisInitialized2["default"])(_this));
    _this.decrementHandler = _this.decrementHandler.bind((0, _assertThisInitialized2["default"])(_this));
    _this.handleButtonsFocus = _this.handleButtonsFocus.bind((0, _assertThisInitialized2["default"])(_this));
    _this.handleChangeTime = _this.handleChangeTime.bind((0, _assertThisInitialized2["default"])(_this));
    _this.handleButtonsDown = _this.handleButtonsDown.bind((0, _assertThisInitialized2["default"])(_this));
    return _this;
  }

  (0, _createClass2["default"])(TimeSelect, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      var prevValue = prevProps.value;
      var value = this.props.value;

      if (prevValue !== value) {
        this.updateTime();
      }
    }
  }, {
    key: "setNextAmPmValue",
    value: function setNextAmPmValue() {
      var ampm = this.state.ampm;
      var nextAmPmValue = (0, _getNextAmPmValue["default"])(ampm);
      this.setState({
        ampm: nextAmPmValue
      });
    }
  }, {
    key: "handleChangeHour",
    value: function handleChangeHour(event) {
      var hour = this.state.hour;
      var value = event.target.value;
      var normalizedValue;

      if ((0, _isNumber["default"])(value)) {
        this.value = value;

        if (Number(value) > 19 || this.isUpOrDownKeyPressed || this.hasPropValue) {
          var newTypedValue = (0, _getSingleNewTypedValue["default"])(hour, value);
          normalizedValue = (0, _normalizeHour["default"])(newTypedValue);
          this.setState({
            hour: normalizedValue
          });
        } else {
          normalizedValue = (0, _normalizeHour["default"])(value);
          this.defaultAmPM = (0, _getDefaultAmPm["default"])(value);
          this.setState({
            hour: normalizedValue
          });
        }

        var shouldNotFocusNextInput = Number(normalizedValue) < 2 && (!hour || this.isUpOrDownKeyPressed || this.hasPropValue);

        if (shouldNotFocusNextInput) {
          this.isUpOrDownKeyPressed = false;
          this.hasPropValue = false;
          return;
        }

        this.isMinutesInputFocused = true;
        this.minutesInputRef.current.focus();
      }
    }
  }, {
    key: "handleFocusHour",
    value: function handleFocusHour() {
      var hour = this.state.hour;
      this.inputFocusedIndex = 0;
      this.prevHour = hour || this.prevHour;
      this.setState({
        hour: ''
      });
    }
  }, {
    key: "handleBlurHour",
    value: function handleBlurHour() {
      var hour = this.state.hour;

      if (this.isUpOrDownButtonPressed) {
        this.isUpOrDownButtonPressed = false;
        return;
      }

      if (this.isMinutesInputFocused) {
        this.isMinutesInputFocused = false;
        return;
      }

      if (hour === '00' && this.value >= '0') {
        this.setState({
          hour: '12'
        });
      }
    }
  }, {
    key: "handleChangeMinutes",
    value: function handleChangeMinutes(event) {
      var minutes = this.state.minutes;
      var value = event.target.value;
      var normalizedValue;

      if ((0, _isNumber["default"])(value)) {
        if (Number(value) > 60 || this.isUpOrDownKeyPressed) {
          var newTypedValue = (0, _getSingleNewTypedValue["default"])(minutes, value);
          normalizedValue = (0, _normalizeMinutes["default"])(newTypedValue);
          this.setState({
            minutes: normalizedValue
          });
        } else {
          normalizedValue = (0, _normalizeMinutes["default"])(value);
          this.setState({
            minutes: normalizedValue
          });
        }

        var shouldNotFocusNextInput = Number(normalizedValue) < 6 && (!minutes || this.isUpOrDownKeyPressed);

        if (shouldNotFocusNextInput) {
          this.isUpOrDownKeyPressed = false;
          return;
        }

        this.amPmInputRef.current.focus();
      }
    }
  }, {
    key: "handleFocusMinutes",
    value: function handleFocusMinutes() {
      var minutes = this.state.minutes;
      this.inputFocusedIndex = 1;
      this.prevMinutes = minutes || this.prevMinutes;
      this.setState({
        minutes: ''
      });
    }
  }, {
    key: "handleAmPmChange",
    value: function handleAmPmChange(value) {
      this.setState({
        ampm: value
      });
    }
  }, {
    key: "hanldeFocusAmPm",
    value: function hanldeFocusAmPm() {
      this.inputFocusedIndex = 2;
    }
  }, {
    key: "handleKeyDown",
    value: function handleKeyDown(event) {
      var keyCode = event.keyCode;

      if (keyCode === _constants.RIGHT_KEY) {
        this.handleRightKeyPressed();
      }

      if (keyCode === _constants.LEFT_KEY) {
        this.handleLeftKeyPressed();
      }

      if (keyCode === _constants.UP_KEY) {
        this.incrementHandler();
      }

      if (keyCode === _constants.DOWN_KEY) {
        this.decrementHandler();
      }

      if (keyCode === _constants.DELETE_KEY) {
        this.resetState();
      }

      if (keyCode === _constants.ENTER_KEY) {
        this.handleChangeTime(event);
      }
    }
  }, {
    key: "handleRightKeyPressed",
    value: function handleRightKeyPressed() {
      var nextInputIndex = this.inputFocusedIndex + 1;
      var nextInputToFocus = this.inputsMap[nextInputIndex];

      if (nextInputToFocus) {
        this.inputFocusedIndex += 1;
        nextInputToFocus.current.focus();
      }
    }
  }, {
    key: "handleLeftKeyPressed",
    value: function handleLeftKeyPressed() {
      var prevInputIndex = this.inputFocusedIndex - 1;
      var prevInputToFocus = this.inputsMap[prevInputIndex];

      if (prevInputToFocus) {
        this.inputFocusedIndex -= 1;
        prevInputToFocus.current.focus();
      }
    }
  }, {
    key: "incrementHandler",
    value: function incrementHandler() {
      this.isUpOrDownKeyPressed = true;

      if (this.inputFocusedIndex === 0) {
        this.incrementHour();
      }

      if (this.inputFocusedIndex === 1) {
        this.incrementMinutes();
      }

      if (this.inputFocusedIndex === 2) {
        this.setNextAmPmValue();
      }
    }
  }, {
    key: "decrementHandler",
    value: function decrementHandler() {
      this.isUpOrDownKeyPressed = true;

      if (this.inputFocusedIndex === 0) {
        this.decrementHour();
      }

      if (this.inputFocusedIndex === 1) {
        this.decrementMinutes();
      }

      if (this.inputFocusedIndex === 2) {
        this.setNextAmPmValue();
      }
    }
  }, {
    key: "handleButtonsDown",
    value: function handleButtonsDown() {
      this.isUpOrDownButtonPressed = true;
    }
  }, {
    key: "handleButtonsFocus",
    value: function handleButtonsFocus() {
      var currentInputFocused = this.inputsMap[this.inputFocusedIndex];

      if (currentInputFocused) {
        currentInputFocused.current.focus();
      }
    }
  }, {
    key: "resetState",
    value: function resetState() {
      if (this.inputFocusedIndex === 0) {
        this.setState({
          hour: ''
        });
        this.prevHour = '';
      }

      if (this.inputFocusedIndex === 1) {
        this.setState({
          minutes: ''
        });
        this.prevMinutes = '';
      }
    }
  }, {
    key: "updateTime",
    value: function updateTime() {
      var value = this.props.value;
      this.setState({
        hour: (0, _getHour["default"])(value),
        minutes: (0, _getMinutes["default"])(value),
        ampm: (0, _getAmPm["default"])(value)
      });
    }
  }, {
    key: "focusHourInput",
    value: function focusHourInput() {
      this.hourInputRef.current.focus();
      this.inputFocusedIndex = 0;
    }
  }, {
    key: "incrementHour",
    value: function incrementHour() {
      var hour = this.state.hour;
      var hourValue = hour || this.prevHour;
      this.setState({
        hour: (0, _normalizeHour["default"])((0, _getNextHour["default"])(hourValue))
      });
    }
  }, {
    key: "decrementHour",
    value: function decrementHour() {
      var hour = this.state.hour;
      var hourValue = hour || this.prevHour;
      this.setState({
        hour: (0, _normalizeHour["default"])((0, _getPrevHour["default"])(hourValue))
      });
    }
  }, {
    key: "incrementMinutes",
    value: function incrementMinutes() {
      var minutes = this.state.minutes;
      var minutesValue = minutes || this.prevMinutes;
      this.setState({
        minutes: (0, _normalizeMinutes["default"])((0, _getNextMinute["default"])(minutesValue))
      });
    }
  }, {
    key: "decrementMinutes",
    value: function decrementMinutes() {
      var minutes = this.state.minutes;
      var minutesValue = minutes || this.prevMinutes;
      this.setState({
        minutes: (0, _normalizeMinutes["default"])((0, _getPrevMinute["default"])(minutesValue))
      });
    }
  }, {
    key: "handleChangeTime",
    value: function handleChangeTime(event) {
      event.preventDefault();
      event.stopPropagation();
      var _this$state = this.state,
          hour = _this$state.hour,
          minutes = _this$state.minutes,
          ampm = _this$state.ampm;
      var _this$props = this.props,
          onChange = _this$props.onChange,
          onCloseModal = _this$props.onCloseModal;
      var currentHour = hour || this.prevHour;
      var currentMinutes = minutes || this.prevMinutes;
      var time = (0, _get24HourTime["default"])({
        hour: currentHour,
        minutes: currentMinutes,
        ampm: ampm
      });

      if (currentHour && currentMinutes && ampm) {
        onChange(time);
      }

      onCloseModal();
    }
  }, {
    key: "render",
    value: function render() {
      var _this$state2 = this.state,
          hour = _this$state2.hour,
          minutes = _this$state2.minutes,
          ampm = _this$state2.ampm;
      var _this$props2 = this.props,
          onCloseModal = _this$props2.onCloseModal,
          cancelLabel = _this$props2.cancelLabel,
          okLabel = _this$props2.okLabel,
          hours24 = _this$props2.hours24,
          className = _this$props2.className;
      var hourPlaceholder = this.prevHour || '--';
      var minutesPlaceholder = this.prevMinutes || '--';
      return _react["default"].createElement("article", {
        className: className
      }, _react["default"].createElement("div", {
        role: "presentation",
        className: "rainbow-time-picker_time-select-content",
        onKeyDown: this.handleKeyDown
      }, _react["default"].createElement("input", {
        "aria-label": "hour",
        onDrop: preventDefault,
        onPaste: preventDefault,
        "data-id": "hour",
        className: "rainbow-time-picker_time-select-value",
        type: "text",
        value: hour,
        placeholder: hourPlaceholder,
        onChange: this.handleChangeHour,
        onFocus: this.handleFocusHour,
        onBlur: this.handleBlurHour,
        inputMode: "numeric",
        pattern: "\\d*",
        ref: this.hourInputRef
      }), _react["default"].createElement("span", {
        className: "rainbow-time-picker_dots"
      }, ":"), _react["default"].createElement("input", {
        "aria-label": "minutes",
        onDrop: preventDefault,
        onPaste: preventDefault,
        "data-id": "minutes",
        className: "rainbow-time-picker_time-select-value",
        tabIndex: "-1",
        type: "text",
        value: minutes,
        placeholder: minutesPlaceholder,
        onChange: this.handleChangeMinutes,
        onFocus: this.handleFocusMinutes,
        inputMode: "numeric",
        pattern: "\\d*",
        ref: this.minutesInputRef
      }), _react["default"].createElement(_RenderIf["default"], {
        isTrue: !hours24
      }, _react["default"].createElement(_ampmSelect["default"], {
        tabIndex: "-1",
        value: ampm,
        defaultValue: this.defaultAmPM,
        onFocus: this.hanldeFocusAmPm,
        onChange: this.handleAmPmChange,
        ref: this.amPmInputRef
      })), _react["default"].createElement("div", {
        className: "rainbow-time-picker_icon-button-container"
      }, _react["default"].createElement(_ButtonIcon["default"], {
        id: "time-picker_up-button",
        tabIndex: "-1",
        variant: "border-filled",
        icon: _react["default"].createElement(_upArrow["default"], null),
        size: "small",
        onMouseDown: this.handleButtonsDown,
        onClick: this.incrementHandler,
        onFocus: this.handleButtonsFocus,
        assistiveText: "Next value"
      }), _react["default"].createElement(_ButtonIcon["default"], {
        id: "time-picker_down-button",
        tabIndex: "-1",
        variant: "border-filled",
        icon: _react["default"].createElement(_downArrow["default"], null),
        size: "small",
        onMouseDown: this.handleButtonsDown,
        onClick: this.decrementHandler,
        onFocus: this.handleButtonsFocus,
        assistiveText: "Previous value"
      }))), _react["default"].createElement("footer", {
        className: "rainbow-time-picker_footer"
      }, _react["default"].createElement(_Button["default"], {
        id: "time-picker_cancel-button",
        className: "rainbow-time-picker_button",
        variant: "base",
        label: cancelLabel,
        onClick: onCloseModal
      }), _react["default"].createElement(_Button["default"], {
        id: "time-picker_ok-button",
        className: "rainbow-time-picker_button",
        variant: "brand",
        label: okLabel,
        onClick: this.handleChangeTime
      })));
    }
  }]);
  return TimeSelect;
}(_react.Component);

exports["default"] = TimeSelect;
TimeSelect.propTypes = {
  onCloseModal: _propTypes["default"].func,
  hours24: _propTypes["default"].bool,
  cancelLabel: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),
  okLabel: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),
  onChange: _propTypes["default"].func,
  value: _propTypes["default"].string,
  className: _propTypes["default"].string
};
TimeSelect.defaultProps = {
  onCloseModal: function onCloseModal() {},
  hours24: false,
  cancelLabel: 'Cancel',
  okLabel: 'OK',
  onChange: function onChange() {},
  value: undefined,
  className: undefined
};
