// react-rainbow-components/components/Table/head/InputCheckbox/label.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Label;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

function Label(props) {
  var label = props.label,
      inputId = props.inputId,
      disabled = props.disabled,
      hideLabel = props.hideLabel,
      id = props.id;

  var getLabelClassNames = function getLabelClassNames() {
    return (0, _classnames["default"])('rainbow-table-input-checkbox_label', {
      'rainbow-table-input-checkbox_label--disabled': disabled,
      'rainbow-table-input-checkbox_label--hide': hideLabel
    });
  };

  return _react["default"].createElement("label", {
    className: "rainbow-table-input-checkbox_label-container",
    htmlFor: inputId,
    id: id
  }, _react["default"].createElement("span", {
    className: "rainbow-table-input-checkbox_faux"
  }), _react["default"].createElement("span", {
    className: getLabelClassNames()
  }, label));
}

Label.propTypes = {
  label: _propTypes["default"].node.isRequired,
  inputId: _propTypes["default"].string.isRequired,
  disabled: _propTypes["default"].bool.isRequired,
  hideLabel: _propTypes["default"].bool,
  id: _propTypes["default"].string
};
Label.defaultProps = {
  hideLabel: false,
  id: undefined
};
