// react-rainbow-components/components/VisualPickerOptionFooter/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = VisualPickerOptionFooter;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _RenderIf = _interopRequireDefault(require("../RenderIf"));

require("./styles.css");

/**
 * A VisualPickerOptionFooter.
 * @category Form
 */
function VisualPickerOptionFooter(props) {
  var label = props.label,
      description = props.description,
      className = props.className,
      style = props.style;

  var getContainerClassName = function getContainerClassName() {
    return (0, _classnames["default"])('rainbow-visual-picker-option-footer_container', className);
  };

  return _react["default"].createElement("article", {
    className: getContainerClassName(),
    style: style
  }, _react["default"].createElement(_RenderIf["default"], {
    isTrue: !!label
  }, _react["default"].createElement("h1", {
    className: "rainbow-visual-picker-option-footer_label"
  }, label)), _react["default"].createElement(_RenderIf["default"], {
    isTrue: !!description
  }, _react["default"].createElement("h2", {
    className: "rainbow-visual-picker-option-footer_description"
  }, description)));
}

VisualPickerOptionFooter.propTypes = {
  /** The label displayed for the VisualPickerOptionFooter. */
  label: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),

  /** The description displayed for the VisualPickerOptionFooter. */
  description: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),

  /** The class name of the root element. */
  className: _propTypes["default"].string,

  /** It is an object with custom style applied to the root element. */
  style: _propTypes["default"].object
};
VisualPickerOptionFooter.defaultProps = {
  label: '',
  description: '',
  className: undefined,
  style: undefined
};
