// react-rainbow-components/components/VerticalSectionOverflow/description.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Description;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _RenderIf = _interopRequireDefault(require("./../RenderIf"));

function Description(_ref) {
  var description = _ref.description,
      isExpanded = _ref.isExpanded;

  var getClassNames = function getClassNames() {
    return (0, _classnames["default"])('rainbow-vertical-section-overflow_action-description', {
      'rainbow-vertical-section-overflow_action-description--expanded': isExpanded
    });
  };

  return _react["default"].createElement(_RenderIf["default"], {
    isTrue: !!description
  }, _react["default"].createElement("span", {
    className: getClassNames()
  }, description));
}

Description.propTypes = {
  description: _propTypes["default"].node,
  isExpanded: _propTypes["default"].bool.isRequired
};
Description.defaultProps = {
  description: ''
};
