// react-rainbow-components/components/Calendar/pageObject/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

/**
 * Calendar page object class.
 * @class
 * @tutorial calendar
 */
var PageCalendar =
/*#__PURE__*/
function () {
  /**
   * Create a new PageCalendar page object.
   * @constructor
   * @param {string} rootElement - The selector of the PageCalendar root element.
   */
  function PageCalendar(rootElement) {
    (0, _classCallCheck2["default"])(this, PageCalendar);
    this.rootElement = rootElement;
  }
  /**
   * Clicks the previous month button element.
   * @method
   */


  (0, _createClass2["default"])(PageCalendar, [{
    key: "clickPrevMonthButton",
    value: function clickPrevMonthButton() {
      $(this.rootElement).$$('button[data-id=button-icon-element]')[0].click();
    }
    /**
     * Clicks the next month button element.
     * @method
     */

  }, {
    key: "clickNextMonthButton",
    value: function clickNextMonthButton() {
      $(this.rootElement).$$('button[data-id=button-icon-element]')[1].click();
    }
    /**
     * Clicks the select year element.
     * @method
     */

  }, {
    key: "clickSelectYear",
    value: function clickSelectYear() {
      $(this.rootElement).$('select').click();
    }
    /**
     * Clicks the specific day button element.
     * @method
     */

  }, {
    key: "clickDay",
    value: function clickDay(day) {
      var buttonEl = $(this.rootElement).$('table').$$('button')[day - 1];
      buttonEl.click();
    }
    /**
     * Returns the text of the current selected month element.
     * @method
     * @returns {string}
     */

  }, {
    key: "getSelectedMonth",
    value: function getSelectedMonth() {
      return $(this.rootElement).$('h3[data-id=month]').getText();
    }
    /**
     * Returns the value of the select year element.
     * @method
     * @returns {string}
     */

  }, {
    key: "getSelectedYear",
    value: function getSelectedYear() {
      return $(this.rootElement).$('select').getValue();
    }
    /**
     * Returns the text of the current selected day element.
     * @method
     * @returns {string}
     */

  }, {
    key: "getSelectedDay",
    value: function getSelectedDay() {
      return $(this.rootElement).$('button[data-selected=true]').getText();
    }
  }]);
  return PageCalendar;
}();

module.exports = PageCalendar;
