// react-rainbow-components/components/CheckboxGroup/checkbox.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _utils = require("../../libs/utils");

var Checkbox =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(Checkbox, _Component);

  function Checkbox(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, Checkbox);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Checkbox).call(this, props));
    _this.inputId = (0, _utils.uniqueId)('checkboxInput');
    return _this;
  }

  (0, _createClass2["default"])(Checkbox, [{
    key: "getLabelClassNames",
    value: function getLabelClassNames() {
      var disabled = this.props.disabled;
      return (0, _classnames["default"])('rainbow-checkbox-group_checkbox-label', {
        'rainbow-checkbox-group_checkbox-label--disabled': disabled
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          value = _this$props.value,
          label = _this$props.label,
          disabled = _this$props.disabled,
          isSelected = _this$props.isSelected,
          onChange = _this$props.onChange,
          describedBy = _this$props.describedBy,
          name = _this$props.name;
      return _react["default"].createElement("div", {
        className: "rainbow-checkbox-group_checkbox"
      }, _react["default"].createElement("input", {
        id: this.inputId,
        type: "checkbox",
        value: value,
        checked: isSelected,
        disabled: disabled,
        onChange: onChange,
        "aria-describedby": describedBy,
        name: name
      }), _react["default"].createElement("label", {
        className: "rainbow-checkbox-group_checkbox-label-container",
        htmlFor: this.inputId
      }, _react["default"].createElement("span", {
        className: "rainbow-checkbox-group_checkbox-faux"
      }), _react["default"].createElement("span", {
        className: this.getLabelClassNames()
      }, label)));
    }
  }]);
  return Checkbox;
}(_react.Component);

exports["default"] = Checkbox;
Checkbox.propTypes = {
  value: _propTypes["default"].string,
  label: _propTypes["default"].node,
  disabled: _propTypes["default"].bool,
  isSelected: _propTypes["default"].bool,
  onChange: _propTypes["default"].func,
  describedBy: _propTypes["default"].string,
  name: _propTypes["default"].string
};
Checkbox.defaultProps = {
  value: undefined,
  label: undefined,
  disabled: false,
  isSelected: false,
  onChange: function onChange() {},
  describedBy: undefined,
  name: undefined
};
