// react-rainbow-components/components/CheckboxGroup/checkboxList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = CheckboxList;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _checkbox = _interopRequireDefault(require("./checkbox"));

var _isOptionSelected = _interopRequireDefault(require("./isOptionSelected"));

function CheckboxList(_ref) {
  var values = _ref.values,
      options = _ref.options,
      onChange = _ref.onChange,
      describedBy = _ref.describedBy,
      name = _ref.name;
  return options.map(function (option, index) {
    var key = "checkbox-".concat(index);
    return _react["default"].createElement(_checkbox["default"], (0, _extends2["default"])({}, option, {
      isSelected: (0, _isOptionSelected["default"])(values, option),
      onChange: onChange,
      describedBy: describedBy,
      key: key,
      name: name
    }));
  });
}

CheckboxList.propTypes = {
  name: _propTypes["default"].string,
  options: _propTypes["default"].array.isRequired,
  values: _propTypes["default"].array,
  onChange: _propTypes["default"].func,
  describedBy: _propTypes["default"].string
};
CheckboxList.defaultProps = {
  values: [],
  onChange: function onChange() {},
  describedBy: undefined,
  name: undefined
};
